### DISCUSSION TITLE
How to speed up updating source cache on background thread?

### DISCUSSION THREAD

**Discussion by Snailya** [reactivemarbles/DynamicData#900] [Q&A] [answered]
In my View, there is a DataGrid bind to a collection in ViewModel. This collection is tranformed from a source cache in a service.
What I suppose is when the view is activated, load the items to list while showing an loading indicator. This means the load task show be execute in a background thread. 
But I found that if I update a souce cache on a background thread, it cosumes much more time than on main thread. For my cirumstance, about 300 item takes 10 seconds on main thread but 2 and a half minutes in background thread.

So is there any way to speed up updating?
By the way, I'm not sure whether it is a common practice for loading items after View show.

**Comment by JakenVeina** [comment]
10 seconds for 300 items is absurd, no matter what thread you're working on, you've got something more serious going on than just a "how do I make this faster" issue.

Can we see some code? If that's not possible or feasible, we can start with some relevant questions...

Where are these items coming from?

Are you adding them one-at-a-time, or in a single batch? Or maybe a handful of batches?

How is this collection exposed to the View layer? An `ObservableCollection<>` generated from the DynamicData `.Bind()` operator?

Are you doing any kind of locking or thread-signaling to access the `SourceCache<>`, or are you just leaving thread-safety to the `SourceCache<>` itself?

> I'm not sure whether it is a common practice for loading items after View show.

That's basically the ideal practice. A) Make sure that users are given visual feedback when data is loading, and B) make sure that the user cannot interact with any UI elements that depend on that data, until it's loaded.

**Reply by Snailya** [reply]
I'm sorry for replying now because I was on vocation.

I'm writing an VSTO addin for Visio. My view is to show the meaningful shapes in the visio drawing. So when the view load, a service is called to get all shapes from the Visio drawing and filter out what is defined to be meaningful then transform them into a models. The model is a also a ReactiveObject as it should response to the change from Visio while the view is showing. And this is done by observe the event from Visio.

The model is transformed first, then added to a source cache using Edit method. The transfromation cost little time, but the edit cost much.

Yes, The source cache is exposed to the View an `ObservableCollection<>` generated from then `.Bind()` operator .

I'm just leaving it to source cache itself. 

I should provide you some code please give me some time. While as you metioned that 10 second is absurd, I think it might be something related to my Reactive Model, I'll check it first.

**Reply by JakenVeina** [reply]
Right on, I'll be happy to look at whatever you can toss up.

**Reply by dwcullop** [reply]
Yes, please, code would be helpful here.

**Reply by Snailya** [reply, marked as answer]
Sorry for the late response due to other works, I think I've figured out recently.
It is not revelant to Dynamic Data, but the COM STA limit for Office. Due to the STA, it has to read data from visio one by one, so it costs much more time than expected even all operation is invoked in the VSTO_Main thread. And it can't be worked around by parallel in background thread, because it will marshal all data to the COM automatically.
Thanks for your help.

**Reply by JakenVeina** [reply]
[MFW COM Interop](https://miro.medium.com/v2/0*ZjYSm_q36J4KChdn).
